### PR TITLE
Add colors that cover 1% of image to palette.

### PR DESCRIPTION
**Pull request by iulia-m-comsa**
Example results: https://docs.google.com/document/d/1mdvt5Q7987wfFwLypDFx7CslQMWmx7R8ytJKSMKKD80/edit?usp=sharing

### CODE REVIEW

File: lib/jxl/modular/transform/palette.h [hunk 1]
```diff
@@ -445,6 +446,44 @@ static Status FwdPalette(Image &input, uint32_t begin_c, uint32_t end_c,
...
+        for (int i = 0; i < 4; ++i) {
+          int dx = offsets[i][0];
+          int dy = offsets[i][1];
+          for (uint32_t c = 0; c < nb; c++) {
+            if (input.channel[begin_c + c].Row(y + dy)[x + dx] != color[c]) {
+              goto out_of_loops;
+            }
+          }
+        }
+        color_freq_map[color] = color_freq_map.find(color) == color_freq_map.end() ? 1 : color_freq_map[color] + 1;
```
Comment (veluca93) [inline lib/jxl/modular/transform/palette.h:471]: `color_freq_map[color] += 1;` works and is faster

**Comment by jonsneyers** [comment]
Do you have some benchmark results for this?
It certainly sounds like a good idea.
Maybe doing the criterion a bit more fuzzy than 1% exact matches (doing some clustering of nearby colors, for a example) might be good to make it more generically applicable.

File: lib/jxl/modular/transform/palette.h [hunk 2]
```diff
@@ -445,6 +446,44 @@ static Status FwdPalette(Image &input, uint32_t begin_c, uint32_t end_c,
...
+        for (uint32_t c = 0; c < nb; c++) {
+          color[c] = p_in[c][x];
+        }
+        int offsets[4][2] = {{1, 0}, {-1, 0}, {0, 1}, {0, -1}};
+        for (int i = 0; i < 4; ++i) {
+          int dx = offsets[i][0];
+          int dy = offsets[i][1];
+          for (uint32_t c = 0; c < nb; c++) {
+            if (input.channel[begin_c + c].Row(y + dy)[x + dx] != color[c]) {
+              goto out_of_loops;
```
Comment (veluca93) [inline lib/jxl/modular/transform/palette.h:467]: I don't see a strong reason for a goto here, boolean variable seems better :D

File: lib/jxl/modular/transform/palette.h [hunk 3]
```diff
@@ -445,6 +446,44 @@ static Status FwdPalette(Image &input, uint32_t begin_c, uint32_t end_c,
   std::vector<pixel_type> color(nb);
   std::vector<float> color_with_error(nb);
   std::vector<const pixel_type *> p_in(nb);
+
+  if(lossy) {
+    // Map colors to their frequencies in the image.
```
Comment (veluca93) [inline lib/jxl/modular/transform/palette.h:451]: This comment doesn't really seem to match what is going on here (I believe we're only counting colors that have an equal 1-cross around them?)

**Review by veluca93 — COMMENTED** [review]
+1 on adding some benchmark results to the CL description

File: lib/jxl/modular/transform/palette.h [hunk 4]
```diff
@@ -445,6 +446,43 @@ static Status FwdPalette(Image &input, uint32_t begin_c, uint32_t end_c,
...
+            if (input.channel[begin_c + c].Row(y + dy)[x + dx] != color[c]) {
+              makes_cross = false;
+            }
+          }
+        }
+        if (makes_cross) color_freq_map[color] += 1;
+      }
+    }
+    // Add colors satisfying frequency condition to the palette.
+    constexpr float image_fraction = 0.01f;
```
Comment (veluca93) [inline lib/jxl/modular/transform/palette.h:476]: kImageFraction

**Review by veluca93 — APPROVED** [review]

File: .gitignore [hunk 5]
```diff
@@ -14,3 +14,10 @@ third_party/hdrvdp-2.2.2.zip.tmp
 # Output plots
 tools/benchmark/metrics/plots
 tools/benchmark/metrics/results.csv
+
+# other
+lib/jxl/modular/transform/.vscode/settings.json
+.gitignore
+lib/jxl/modular/transform/else if (use_kmeans) {.cpp
```
Comment (veluca93) [inline .gitignore:21]: This and .gitignore should not be here

File: lib/jxl/modular/transform/palette.h (hunk 1 above)
Comment (iulia-m-comsa) [reply lib/jxl/modular/transform/palette.h]: Done.

File: lib/jxl/modular/transform/palette.h (hunk 2 above)
Comment (iulia-m-comsa) [reply lib/jxl/modular/transform/palette.h]: Done.

**Review by veluca93 — APPROVED** [review]

File: lib/jxl/modular/transform/palette.h (hunk 3 above)
Comment (iulia-m-comsa) [reply lib/jxl/modular/transform/palette.h]: Changed the comment.

File: lib/jxl/modular/transform/palette.h (hunk 4 above)
Comment (iulia-m-comsa) [reply lib/jxl/modular/transform/palette.h]: Done.

File: .gitignore (hunk 5 above)
Comment (iulia-m-comsa) [reply .gitignore]: It was not supposed to :D

**Review by veluca93 — APPROVED** [review]

**Pull request merged**